### ISSUE TITLE
Optional type not checked if used in only one branch of nested promises

### ISSUE THREAD

**Issue by SimonSimCity** [microsoft/TypeScript#42417]
# Bug Report

The system doesn't report the wrong type of an optional property if only one branch returns the optional property, but another branch skips it.

Another ticket I posted recently has a very similar scenario but falsely reports on correct code (whereby this is the opposite - it falsely passes invalid code): #42293

### 🔎 Search Terms

promise, catch, optional property, wrong type

### 🕗 Version & Regression Information

* 3.5.1
* 4.1.3
* 4.2.0-beta
* nightly (4.2.0-dev.20210120)

### ⏯ Playground Link

This playground shows 3 examples of the code posted below. The second two examples are minor changes which trigger the expected behavior:

https://www.typescriptlang.org/play?ts=4.1.3#code/KYDwDg9gTgLgBDAnmYcAiBDGG4F44AKUEAtgJYDOwAPAN4BQcTcUwAjgK5msAmREKWIgBccAHYcSAI2BQA3I2YCYZCGIwAbfoKQBGAPyiJ02QoC+APgVN6AYzUV4ANwxQyGKRuCjM2PIWJyKgA6VgoIDSdgAAoASmCYAAtgMWi4vAs4BiYyADM4aIBZLETQjDEeUnSAKjgAJjhM3VisxSZWGA4oMVbmZlZObmA+Yh0ROAAGABo2pTAVNU1tWT1RACIPWzWZvrNrODN6RQ6unv4g4FDgcMiY+KSUtJbcTOz+4E7u3r6Wdi5eZZCUTTWZ7RRmWLmSFHAD0MLgAGFoKxbDANIhfpBYBQ4GR4BgcbJiFA4AB3ZI9HijMBkMQAcwQyTgVHsFTgYEClGAAFpSVAMGAafS7A5nK53J5gHUfFgcPhzlyrjconEEhSnhlvnkCsUkmUKlUWrUGk0Wm92h9Tt8+gN-sNAUhgTsfspVOotKMVohdOtNttQftDoo4b9PmdOSEwhEVfd1ekXtbQ1bze9BgDPUDJs6mGCmCGIVCFPQQ0ioCi0RjWFiYDi8XACXAidAyRS4BwKLSGQ84K7Fh6BF7dD2elIIEkk91uWAShQRWJHHAXG4PF4AMwyvzyiOXKO3VUPVLx16KbVFEr6yokGr1RpwZqJk5fFMWtP2jOOrOzJi990O72+qQtmzA5AyOC0wwCUhFV3GM1UeI8H0tJ8v1+V8RgHTMQR+Ht5jdJZ33-OANkA-1dkDaECyOVkbkuDQIDpaIlwlLwpkXcUVylVimI41dYnoIA

### 💻 Code

```ts
export type Data = Promise<{
    requiredProperty: number;
    optionalProperty1?: number;
}>;
  
const variable: Data = Promise.resolve().then(() => {
  if (Math.random() * 2 > 1) {
    return {
      requiredProperty: 0,
      optionalProperty1: "abc",
    };
  }

  return Promise.resolve().then(() => {
    return {
      requiredProperty: 0,
    };
  });
});
```

### 🙁 Actual behavior

No error reported

### 🙂 Expected behavior

The linter should complain by telling:

```ts
Type 'Promise<{ requiredProperty: number; optionalProperty1?: number | undefined; } | { requiredProperty: number; optionalProperty1: string; }>' is not assignable to type 'Data'.
  Type '{ requiredProperty: number; optionalProperty1?: number | undefined; } | { requiredProperty: number; optionalProperty1: string; }' is not assignable to type '{ requiredProperty: number; optionalProperty1?: number | undefined; }'.
    Type '{ requiredProperty: number; optionalProperty1: string; }' is not assignable to type '{ requiredProperty: number; optionalProperty1?: number | undefined; }'.
      Types of property 'optionalProperty1' are incompatible.
        Type 'string' is not assignable to type 'number | undefined'.
```

**Comment by RyanCavanaugh** [comment]
There's a subtype reduction step that occurs here that we need to make this check correctly, and the optional property can get aliased away in that step. In general you can will always be able to write *some* code that will encounter an unsoundness with optional properties due to object types not being sealed; this is one such example.

**Comment by acutmore** [comment]
Interestingly the error _is_ caught when the `.then` callback is an async function.

https://www.typescriptlang.org/play?ts=4.1.3#code/C4TwDgpgBAIghsOUC8UAKAnA9gWwJYDOEAPAN4BQUUGEAjgK540AmmWkGoAXFAHb04ARhAwAaSlHbA8WXnAA2bDqACMAfh78hI8QF8AfAG5y5OARC8AxlABm9K9NlRgEAsAAUASigUqAej8qIOCQ0LDwiIA3aOjqCBwsSOgAcjMLS2SoOF5mKBFsDAIoZkICOEg4QolLWTcoSMq8OEF5CB54RBR0bHwiADoaAix5JK8+4AALCF53NKsvFH0fCSo8Gyh3AFkECYHs5lwFgCooACYoJZVvXxCaYHoMXmXwmgYmCFZsZRAeAAZxcJSGRyRRfESqHgAImalkhAJCumMwV0JmCdweTzYvQgA1cw1GnnGUxmC2QSxuaIg90ezzCr0YLCU4J+UH+KyCiPZuk8SKg3OMKKAA
